src/Ganeti/HTools/Backend/IAlloc.hs

```diff
@@ -233,8 +234,19 @@ parseData now body = do
       dsrd_loc_tags = extractDesiredLocations ctags
       updateTags = updateExclTags ex_tags .
                    updateDesiredLocationTags dsrd_loc_tags
-  cdata1 <- mergeData [] [] [] [] now (ClusterData gl nl il ctags defIPolicy)
-  let (msgs, fix_nl) = checkData (cdNodes cdata1) (cdInstances cdata1)
+  -- hypervisor
+  enabled_hypervisors <- extrObj "enabled_hypervisors"
+  -- This is ugly, unfortunately the  IAllocator "API" (which is a CLI not an
+  -- API to be precise) is different from other htools backends that receive
+  -- the 'default_hypervisor' field. Normally there is no reason to have
+  -- more than one hypervisor so in practice this should be fine. Ganeti 2.17
+  -- exports the hv_stat per node field with the missing information.
+  -- TODO: Remove, once the caller can get the correct nodes size on KVM.
+  let nl2 = case enabled_hypervisors of
```

Comment:
I have followed the same pattern as with other 'sane' htools backends setting the hypervisor parameter just before creating the ClusterData record in parseData. The parsing is backend specific so that plugin would be Json backend specific. 

The ugly thing is that IAllocator interface is broken by design. IAllocator takes a json file for input that has limited (different) semantics compared to Htools backends (Rapi/Luxi/Text etc.) While other backends separate parsing and merging the data, IAllocator does it in one step. I think the support for Text backend in IAllocator was carelessly hacked in. First it parses the json  (r1 in Hail.wrapReadRequest) then if optDataFile (Text backend) or optNodeSim (Simu backend) is specified, the json loaded cluster data is thrown away and a new one is loaded using the standard htools backends. 

I think fixing this would require redesigning IAllocator interface and a significant effort that I think should be a separate fix.